### PR TITLE
*: Enable refactored authority discovery

### PR DESCRIPTION
**Pull request by mxinden**
This pull request introduces the [refactored version](https://github.com/paritytech/substrate/pull/3955) of the *authority discovery* module into the `v0.6` *Kusama* branch. I would later on cherry-pick it into `master`.

### Rollout strategy

The *authority discovery* module has not been tested on a large network so far. I would suggest not to roll it out to all nodes at once, but at the same time I am also not familiar with how we roll out new features in Kusama in general. Thoughts? Is this too careful?

*Authority discovery* requires a runtime module. In addition it depends on other nodes putting their identity onto the DHT. Thus we can't just run a single custom node. We could gate the feature behind a cli flag and enable it on two of our validators once the feature is rolled out.

### Additional required changes due to session key order change

@bkchr mentioned that the UI depends on the order of the session keys as of today. Given that this pull request alters this order @jacogr what are the necessary external changes needed?

### CODE REVIEW

**Comment by jacogr** [comment]
@mxinden I have a PR for that already in the API, e.g. https://github.com/polkadot-js/api/pull/1573

It is a bit tricky since it will need to go live with the Kusama upgrade. (However as soon as merged here, will merge there and then merge the appropriate UI PR as the change goes live on Kusama with a runtime upgrade)

I'm more worried about the validators swapping over than anything else.

File: runtime/src/lib.rs [hunk 1]
```diff
@@ -253,6 +254,7 @@ impl_opaque_keys! {
 		pub grandpa: Grandpa,
 		pub babe: Babe,
 		pub im_online: ImOnline,
+		pub authority_discovery: AuthorityDiscovery,
```
Comment (gavofyork) [inline runtime/src/lib.rs]: should probably be in the line below so the existing key types don't switch their location.

File: service/src/chain_spec.rs [hunk 2]
```diff
@@ -194,6 +203,7 @@ pub fn get_authority_keys_from_seed(seed: &str) -> (
 	BabeId,
 	GrandpaId,
 	ImOnlineId,
+	AuthorityDiscoveryId,
```
Comment (gavofyork) [inline service/src/chain_spec.rs]: same here.

File: service/src/chain_spec.rs [hunk 3]
```diff
@@ -202,13 +212,14 @@ pub fn get_authority_keys_from_seed(seed: &str) -> (
...
 		get_from_seed::<ImOnlineId>(seed),
+		get_from_seed::<AuthorityDiscoveryId>(seed),
 		get_from_seed::<ValidatorId>(seed),
 	)
 }
 
 /// Helper function to create GenesisConfig for testing
 pub fn testnet_genesis(
-	initial_authorities: Vec<(AccountId, AccountId, BabeId, GrandpaId, ImOnlineId, ValidatorId)>,
+	initial_authorities: Vec<(AccountId, AccountId, BabeId, GrandpaId, ImOnlineId, AuthorityDiscoveryId, ValidatorId)>,
```
Comment (gavofyork) [inline service/src/chain_spec.rs]: same here.

File: service/src/chain_spec.rs [hunk 4]
```diff
@@ -202,13 +212,14 @@ pub fn get_authority_keys_from_seed(seed: &str) -> (
 		get_from_seed::<BabeId>(seed),
 		get_from_seed::<GrandpaId>(seed),
 		get_from_seed::<ImOnlineId>(seed),
+		get_from_seed::<AuthorityDiscoveryId>(seed),
```
Comment (gavofyork) [inline service/src/chain_spec.rs]: same here

File: service/src/chain_spec.rs (hunk 2 above)
Comment (gavofyork) [inline service/src/chain_spec.rs]: same here

**Comment by gavofyork** [comment]
looking at the code, i'm pretty sure a special migration will be needed for `NextKeys` and `QueuedKeys` since it'll try to decode the entire tuple, which is now different, all at once. annoyingly, we can't enumerate either.

the easiest way forward will probably just be to create a custom decode function for the `Keys` type which appends zeroes onto the end of the input in case it's too short.

wdyt @bkchr ?

**Comment by bkchr** [comment]
As described in Riot, we can iterate the `Session` `NextKeys` and `QueuedKeys` as both are a double map with a fixed first key. However, your proposed solution with a custom decode could work as well.

**Comment by mxinden** [comment]
To make sure I understand your suggestion @bkchr , I could:

1. Wait for storage to be iteratable for a specific prefix (Related to https://github.com/paritytech/substrate/pull/4185).

2. Within the `on_initialize` of the session module of the next runtime, iterate
   over the keys of the `QueuedKeys` and `NextKeys` prefix, parse each `T::Keys`
   value as if it contains 4 keys (which it still would) and write it back as 5
   keys (4 + authority discovery key) with the last either all zero, or as an
   `Option<Key>::None`.
   
3. Eventually once someone generates a new session key, the 5th key will be
   derived, replacing the 5th key within the storage of either all zero or `None`
   with an actual authority discovery key.

**Comment by gavofyork** [comment]
that's right, though if we want to move things along without blocking on paritytech/substrate#4185  then we can also add a special `Decode` into key tuples inside the `impl_opaque_keys` macro, which makes it zero-pad the input (using e.g. `TrailingZeroInput`) so that if there's only 4 keys explicitly in storage (as there would be) it'll still find a bunch of trailing zeros for the 5th.

**Comment by mxinden** [comment]
Cherrypicking this change set on top of `master` right now.

**Comment by mxinden** [comment]
https://github.com/paritytech/polkadot/pull/601/commits/6355956087b0680e1154c78f7408cf113c7bec08 gates the authority discovery module behind the `--enable-authority-discovery` feature flag.

@bkchr @gavofyork  could you give this another review?

**Comment by mxinden** [comment]
`mxinden:enable-authority-discovery-master` contains the same commits cherry-picked on top of current `paritytech:master`.

https://github.com/paritytech/polkadot/compare/master...mxinden:enable-authority-discovery-master?expand=1

**Comment by mxinden** [comment]
Replaced by https://github.com/paritytech/polkadot/pull/624.

**Pull request closed without merging**